proxy/stats_get_prediction.go

```diff
@@ -0,0 +1,361 @@
+			s.predictionState = reset
+		}
+
+		if path == "/api/sys/get_env" || path == "/api/user/login" {
+			wrappedWriter := CachingResponseWriter{w: w}
+			next.ServeHTTP(&wrappedWriter, r)
+
+			if path == "/api/sys/get_env" && wrappedWriter.code < 400 {
+				s.predictionState = get_env_called
+			} else if path == "/api/user/login" &&
+				wrappedWriter.code < 400 &&
+				s.predictionState == get_env_called {
+				login := make([]byte, wrappedWriter.buf.Len())
+
+				wrappedWriter.buf.Read(login)
+				s.ParseLoginPrefix(login)
+				s.predictionState = login_parsed
+			}
+		} else {
+			next.ServeHTTP(w, r)
+		}
+	})
+
+}
+
+// Proxy getstats
+func (s *StatsGetPrediction) HandleGetStats(w http.ResponseWriter, r *http.Request) bool {
+	if len(s.loginPrefix) > 0 && s.predictionState == login_parsed {
+		s.AsyncGetStats()
+		s.predictionState = sending_calls
+	}
+	if len(s.loginPrefix) > 0 && s.predictionState == sending_calls {
+		bodyBytes, _ := ioutil.ReadAll(r.Body)
+		r.Body.Close() //  must close
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
+		req := string(bodyBytes)
+		if task, ok := s.statsGetTasks[req]; ok {
+			resp := <-task.response
+			if resp == nil {
+				fmt.Println("Cache Error!")
+				delete(s.statsGetTasks, req)
+				return false
+			}
+			defer resp.Body.Close()
+			// Copy headers
+			for name, values := range resp.Header {
+				w.Header()[name] = values
+			}
+			w.WriteHeader(resp.StatusCode)
+			reader := io.TeeReader(resp.Body, w) // For dumping API payloads
+			_, err := io.ReadAll(reader)
+			if err != nil {
+				fmt.Println(err)
+			}
+			delete(s.statsGetTasks, req)
+			return true
+		}
+		fmt.Println("Cache miss! " + req)
+		return false
+	}
+	return false
+}
+
+func (s *StatsGetPrediction) ParseLoginPrefix(loginRet []byte) {
+	s.loginPrefix = hex.EncodeToString(loginRet[60:79]) + hex.EncodeToString(loginRet[2:16])
+}
+
+func (s *StatsGetPrediction) BuildStatsReqBody(login string, req string) string {
+	/*
+
+		Get Stats Call Analysis
+		E.g.
+		data=9295xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx02a5302e302e350396a007ffffffff^@
+
+		1.
+		"data="
+		length=5
+
+		2. Header?
+		9295
+		length=2
+
+
+		3. Login 1?
+		index in login response=60
+		length=19
+
+		4. Login 2?
+		index in login response=2
+		length=14
+
+		5. Divider?
+		02a5
+		l=2
+
+		6. Version?
+		302e302e35 (0.0.5)
+		l=5
+
+		7. Divider2?
+		0396
+		l=2
+
+		8. Specific call
+		e.g. a007ffffffff , Confirm that this stays between users
+		l=6
+
+
+		9=End
+		\0
+		l=1
+	*/
+
+	var sb strings.Builder
+	sb.WriteString("data=")
+	sb.WriteString("9295") // Header
+	sb.WriteString(login)
+	sb.WriteString("02a5")       // Divider
+	sb.WriteString("302e302e35") // 0.0.5
+	sb.WriteString("0396")       // Divider 2
+	sb.WriteString(req)
+	sb.WriteString("\x00") // End
+	return sb.String()
+}
+
+// Process the filled queue, then exit when it's empty
+func (s *StatsGetPrediction) ProcessStatsQueue(queue chan *StatsGetTask) {
```

Comment:
Best practice is to limit the channel direction ie `<-chan` so you can't accidentally insert into the channel from a consumer.